### PR TITLE
feat(admin): organization dashboard

### PR DESCRIPTION
**Pull request by dlnanh85**
Dashboard page for GDSC Admin to modify Organizations.
- Add page /admin/organizations

**Notes:**
- No APIs available to apply yet -> using mock data
- All codes are placed inside /src/app/admin for isolation purpose (as PM requested)

### CODE REVIEW

**Review by NooBat — COMMENTED** [review]
What is the point of using a `src` directory inside here? I think this only makes it harder to actually visualize the structure. Please use the existed directories and refactor svgs by putting it inside the `raw` directory and run `yarn add:img`

**Review by NooBat — COMMENTED** [review]
These two modals are quite similar, maybe make a generic component Modal?

File: src/app/admin/organizations/CreateOrganizationModal.tsx [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+
+import OrganizationForm from './OrganizationForm';
+
+type OrganizationModalProps = {
+  onCreate: (organization: BaseOrganization) => void;
+  onCancel: () => void;
+};
+
+export default function CreateOrganizationModal(props: OrganizationModalProps) {
+  const { onCreate, onCancel } = props;
```
[the remark this answers is not in the document]
Comment (dlnanh85) [reply src/app/admin/organizations/CreateOrganizationModal.tsx:16]: It was a matter of convention instructed by Mr. Đức, I was not questioning at the time, as I couldn't see the difference or harm. I will update the convention as you wish. Thank you.

File: src/app/admin/organizations/CreateOrganizationModal.tsx [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+  const [isLongNameValid, setIsLongNameValid] = useState(true);
+  const longNameContainerRef = useRef<HTMLDivElement>(null);
+
+  return (
+    <div
+      onClick={onCancel}
+      className='fixed bottom-0 left-0 right-0 top-0 z-10 flex items-center justify-center bg-black/20'
+    >
+      <div
+        onClick={(e: MouseEvent) => e.stopPropagation()}
```
[the remark this answers is not in the document]
Comment (dlnanh85) [reply src/app/admin/organizations/CreateOrganizationModal.tsx:34]: The `div` on line 29 acts as a shaded background that fill all the page, and the `div` on line 33 is the actual modal. I want that when the user clicked outside the modal (which is click on the shaded background), the modal will be closed (call the `onCancel` handler).

Therefore, when the user is using the modal, I need to stop the `click` event on the actual modal to bubbling up the background and accidentally close the modal.

So I think it's still appropriate for `div`s to have `onClick`

If you see a better approach, please let me know.

File: src/app/admin/organizations/DomainList.tsx [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+
+type DomainListProps = {
+  domains: string[];
+  onDomainsChange: (domains: string[]) => void;
+};
+
+export default function DomainList(props: DomainListProps) {
+  const { domains, onDomainsChange } = props;
+  const [isAdding, setIsAdding] = useState(false);
+  const [domainInput, setDomainInput] = useState<null | string>(null);
```
[the remark this answers is not in the document]
Comment (dlnanh85) [reply src/app/admin/organizations/DomainList.tsx:21]: I mistook `''` is `truthy`. I will remove the `null`.

File: src/app/admin/organizations/OrganizationForm.tsx [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+                setEditingOrganization({
+                  ...editingOrganization,
+                  managers: users,
+                });
+              }}
+              options={nonManagingUsers}
+              isManagerList
+            />
+          </div>
+          {nonParticipatingUsers && (
```
[the remark this answers is not in the document]
Comment (dlnanh85) [reply src/app/admin/organizations/OrganizationForm.tsx:112]: `nonParticipatingUsers` are filtered from the initialUsers array, which will be acquired through API call, therefore during first render, `nonParticipatingUsers` will be `undefined`.

File: src/app/admin/organizations/OrganizationItem.tsx [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+  index: number;
+  onEdit: () => void;
+};
+
+const MAX_MANAGERS_DISPLAY = 3;
+
+export default function OrganizationItem(props: OrganizationItemProps) {
+  const { organization, index, onEdit } = props;
+
+  const renderManagers = () => {
```
[the remark this answers is not in the document]
Comment (dlnanh85) [reply src/app/admin/organizations/OrganizationItem.tsx:20]: At the time, I wanted to separate the logic for rendering managers, and I immediately create a `render` function right inside the component (It's embarrass, but I don't know why I did it). 

I have separated it into new component.

**Comment by dlnanh85** [comment]
> What is the point of using a `src` directory inside here? I think this only makes it harder to actually visualize the structure. Please use the existed directories and refactor svgs by putting it inside the `raw` directory and run `yarn add:img`

As I mentioned, the Mr. An requested this specifically, because he is considering seperating this admin feature into a separated project, so I want to code this as isolated as possible.

That being said, I am also using resources from outside this src/app/admin folder. This Sunday, we will discuss this further, and decide an appropriate approach.


> These two modals are quite similar, maybe make a generic component Modal?

I was not sure how to approach having modals with the same UI, but different kind of event handler. 

On the one hand, if I put `onSave: (organization: Organization | BaseOrganization) => void` for both create and edit scenario, TypeScript will shout at my face when I try to pass the parameter as `BaseOrganization`, which I understand.

On the other hand, I can separate them into `onCreate` and `onEdit` props, but I have to make both of them optional, which I find not ideal also (I am thinking of the scenario when someone use that component and not passing any handler, and still pass TypeScript)

So my solution is try to condense the duplicate UI into a pure UI component `OrganizationForm`, and let the higher components `CreateOrganizationModal` and `EditOrganizationModal` handling the differences in handling logic.

File: .eslintrc.cjs [hunk 6]
```diff
@@ -315,7 +315,7 @@ module.exports = {
         'max-lines-per-function': [
           'error',
           {
-            max: 150,
+            max: 160,
```
Comment (dlnanh85) [inline .eslintrc.cjs:318]: This is temporary, as I cannot pass lint-stage.

File /src/app/admin/components/OrganizationModal/BaseOrganizationModal.tsx exceeds the 150 lines per function limit

**Pull request closed without merging**